src/Language/Docker/Parser/Copy.hs

```diff
@@ -66,6 +67,15 @@ parseAdd = do
                   c : _ -> c
       fileList "ADD" (\src dest -> Add (AddArgs src dest cho chm))
 
+heredocList :: (?esc :: Char) =>
+               (NonEmpty SourcePath -> TargetPath -> Instruction Text) ->
+               Parser (Instruction Text)
+heredocList constr = do
+  sources <- spaceSep1 heredocMarker
+  target <- untilEol "target path"
+  void $ heredocContent (last sources)
```

Comment:
Even when it is obviously correct here, the use of the partial fiction `last` makes me uncomfortable. Can you use pattern matching and fail if the list is empty?

Reply:
Pattern matching in this case really is not improving all that much, in particular the readability suffers quite a bit IMO. But then again I understand your concerns about `last`.